pyteal/ast/scratchvar.py

```diff
@@ -42,7 +61,12 @@ def store(self, value: Expr) -> Expr:
         Args:
             value: The value to store. Must conform to this ScratchVar's type.
         """
-        require_type(value, self.type)
+        from pyteal.ast.subroutine import SubroutineCall
```

Comment:
I'm trying to understand why we need this workaround for Subroutines. Maybe there exists a more straightforward approach?

Reply:
Let me provide a little more context about the design back when we were working on `ABIReturnSubroutine`.

The `ABIReturnSubroutine` returns the `output` value encoding on the stack, then the `deferred_expr` picks it up and store the encoding into a scratch slot, when we have only scratch slot. Anyways, the stack remains nothing, and surprisingly the `ABIReturnSubroutine` evaluates to nothing, i.e., `TealType.none`.

Now what, what is it related with `BaseType.store_value`? There's some hand waving between `deferred_expr` and `BaseType.store_value`: such that after subroutine call, we initialize an ABI value. The ABI value stores the result of `ABIReturnSubroutine` into itself, which is the execution of `deferred_expr`.

So why it is using `slot` rather than just `store` into `scratchvar`? Recall we discussed that `ABIReturnSubroutine` has type `TealType.none`, we here use `slot` to bypass the type checking, which is not quite good.

Now we are back to this work around, so for subroutine call with abi return, we need to check the type of output, such that we can perform sort of type check against stack

Reply:
Is there a specific test that you know of that is failing without introducing this code?

Assuming we need this code, I think a slight improvement (in terms of legibility) would be to consider the `else` case first.

```python
        if isinstance(value, SubroutineCall) and value.output_kwarg:
            assert types_match(value.output_kwarg.abi_type.storage_type(), self.type)
        else:
            require_type(value, self.type)
```

Reply:
I opened PR: https://github.com/algorand/pyteal/pull/588 to illustrate the results of lifting such checks.

Essentially, on all of build test failing cases: they are attempting to store results of `ABIReturnSubroutine` into another ABI value (with underlying scratch var). The result of `ABIReturnSubroutine` should be of type `ReturnedType`, and the `type_of` method is returning `TealType.none`.

If we want to match the `ScratchVar`'s type against `ReturnedValue`'s "wrapped" result type, we need to go into `value.output_kwarg` to see the type of ABI value encodings.

Hopefully there are more succinct ways to deal with the type checking... 🤔

Reply:
In any case, this piece of code only perform type checking, I should do a better job in raising error message.

Reply:
I'm investigating and close to prepare a proposed child PR.

Reply:
Here is the child PR #590. There are pros and cons to it. I am not convinced that it is actually an improvement over the current approach. However, if you feel that it is feel free to merge it in. Or if you see some possible improvements that would get it mergeable, feel free to suggest. However, given the very modest effect on overall code quality, we shouldn't spend much more time in this rabbit hole which I have dug for us.

Reply:
I am merging in the change, for I think it is a reasonable move for code type check.